thorlcr/activities/firstn/thfirstnslave.cpp

```diff
@@ -263,17 +263,20 @@ class CFirstNSlaveGlobal : public CFirstNSlaveBase, implements ILookAheadStopNot
         totallimit = (rowcount_t)helper->getLimit();
 
         rowcount_t _skipCount = validRC(helper->numToSkip()); // max
-        byte newFastThroughState = (byte)(::isFastThrough(input) ? 1 : 0);
+        byte newFastThroughState = (byte)(isInputFastThrough(0) ? 1 : 0);
         if ((lastFastThroughState != newFastThroughState) || (lastTotalLimitState != totallimit) || (lastSkipCountState != _skipCount))
         {
             if (0 == newFastThroughState)
             {
-                /* JCSMORE: am here because something changed, but could be the fast through state changed backward and forth
-                 * such that there was a previously installed look ahead with the same 'limit' and 'skip' setup, in which case could reuse.
-                 */
-                rowcount_t maxRead = (totallimit>(RCUNBOUND-_skipCount))?RCUNBOUND:totallimit+_skipCount;
-                setLookAhead(0, createRowStreamLookAhead(this, inputStream, queryRowInterfaces(input), FIRSTN_SMART_BUFFER_SIZE, ::canStall(input), false,
-                                                                                  maxRead, this, &container.queryJob().queryIDiskUsage()), false); // if a very large limit don't bother truncating
+                // something changed, but may already have lookahead installed, check limit/skip and if same, just start existing one.
+                if (hasLookAhead(0) && (lastTotalLimitState == totallimit) && (lastSkipCountState == _skipCount))
```

Comment:
Not quite right - lastXXX is the last values when it was started, not the last values for the lookahead.  So lookahead(4), nolookahead(5), lookahead(5) would reuse the lookahead, but with value of 4 instead of 5.

Reply:
It's over complicated, will simplify in next commit in a moment.